### PR TITLE
Staking

### PR DESCRIPTION
**Pull request by tensojka**

### CODE REVIEW

**Comment by tensojka** [comment]
known issue: can unstake when tokens are delegated

**Comment by tensojka** [comment]
also missing: events

**Review by Nerrolol — COMMENTED** [review]
The whole structure looks good to me as well as the code (I may need another look to make sure I got it all).
A few questions : atm the staking curve is set only for points 1M, 3M, 6M, 1Y with pre-defined conversion rates. I assume it's for testing purposes only and a true staking curve logic is to be implemented ?

In the staking_tests file, when you verify the balance of the staker is correct (stake of 937 for 3 months => 2342 with the 2,5X conversion rate) why is it rounded to the lowest int?

Maybe a quick test add of unstaking before the unlock_date

**Comment by tensojka** [comment]
@Nerrolol

The production curve will still be discrete.

It's rounded to the lowest int, because, let's assume we rounded up and not down: let's say you want to lock 1000 KONOHA for a duration that gives you a 1.6x multiplier.

If you lock 1000 directly, you get 1600 veKONOHA.

But if you lock 1000x1 KONOHA (call the function stake() 1000 times), you get 2000 veKONOHA.

Adding a test of unstaking before expiry

**Review by Nerrolol — APPROVED** [review]
lgtm

File: src/token.cairo [hunk 1]
```diff
@@ -1,13 +1,21 @@
 // This is a freely tradable ERC20 token.
-// TODO upgradability
 
 #[starknet::contract]
 mod FloatingToken {
```
Comment (Chepelau) [inline src/token.cairo:4]: Isn't this token missing mint and burn functions?

File: src/staking.cairo [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+            assert(amount != 0, 'amount to stake is zero');
+            let conversion_rate: u16 = self.curve.read(length);
+            assert(conversion_rate != 0, 'unsupported stake length');
+
+            let floating_token = IERC20Dispatcher {
+                contract_address: self.floating_token_address.read()
+            };
+            floating_token.transfer_from(caller, get_contract_address(), amount.into());
+
+            let amount_voting_token = (amount * conversion_rate.into()) / 100;
```
[the remark this answers is not in the document]
Comment (tensojka) [reply src/staking.cairo:134]: This should already throw away the remainder, but better to use divrem i guess

File: src/token.cairo (hunk 1 above)
Comment (tensojka) [reply src/token.cairo:4]: The CARM token has fixed supply, so no mints/ burns

Comment (Chepelau) [reply src/token.cairo:4]: In function `unstake_airdrop` in `staking.cairo` the `mint` function is called on floating_token

Comment (tensojka) [reply src/token.cairo:4]: fixed thx

**Review by Chepelau — APPROVED** [review]

**Review by Chepelau — APPROVED** [review]

**Pull request merged**